File: 1475-maximum-sum-bst-in-binary-tree/maximum-sum-bst-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    struct Node{
        int minVal;
        int maxVal; 
        int sum;
        Node(int minVal,int maxVal,int sum){
            this->minVal = minVal;
            this->maxVal = maxVal;
            this->sum = sum;
        }
    };
    
    Node* isValid(TreeNode* root, int &bstSum){

        if(root==NULL){return new Node(INT_MAX,INT_MIN,0);}

        Node* leftTree = isValid(root->left, bstSum);
        Node*  rightTree = isValid(root->right, bstSum);
        
        int val = root->val;
        
        if(leftTree == NULL || rightTree == NULL || val <= leftTree->maxVal || val >= rightTree->minVal){
            return NULL;
        }
        // for given root value it should between max value in leftsubtree and min value in right subtree

        int sum = val + leftTree->sum + rightTree->sum;
        bstSum = max(sum,bstSum);
        
        int mi = min(val, leftTree->minVal);// leftTree[0] has the minimum value in left sub tree
        int ma = max(val, rightTree->maxVal);// rightTree[1] has the maximum value in right sub tree

        return new Node(mi,ma,sum);
    }

    int maxSumBST(TreeNode* root) {
        int bstSum = 0;
        isValid(root,bstSum);
        return bstSum;
        // bottom up solution
    }
};

File: 1475-maximum-sum-bst-in-binary-tree/maximum-sum-bst-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int[] findMaxSum(TreeNode root, int[] ans){
        int[] res = new int[4];
        if(root==null){
            res[0]=Integer.MAX_VALUE;
            res[1]=Integer.MIN_VALUE;
            return res;
        }

        int[] left = findMaxSum(root.left,ans);
        int[] right = findMaxSum(root.right,ans);

        if(left[3] == 1 || right[3] ==1 || root.val<=left[1] || root.val>=right[0]){
            res[3]=1;
            return res;
        }

        int sum = root.val;
        sum += left[2] + right[2];
        ans[0] = Math.max(sum,ans[0]);
        res[0]=Math.min(root.val,left[0]);
        res[1]=Math.max(root.val,right[1]);
        res[2] = sum;
        return res;
    }

    public int maxSumBST(TreeNode root) {
        int[] ans = new int[1];
        findMaxSum(root,ans);
        return ans[0];
    }
}